reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+
+    public AnswerCountEstimator(LogicManager logicMgr) {
+        this.logicMgr = logicMgr;
+        this.answerCountModel = new AnswerCountModel(this);
+        this.estimators = new HashMap<>();
+        this.initializationStack = new ArrayList<>();
+        this.cycleHeads = new HashSet<>();
+        this.toReset = new HashSet<>();
+    }
+
+    public long estimateAllAnswers(ResolvableConjunction conjunction) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAllAnswers();
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter) {
+        return estimateAnswers(conjunction, variableFilter, logicMgr.compile(conjunction));
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter, Set<Resolvable<?>> includedResolvables) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAnswers(variableFilter, includedResolvables);
+    }
+
+    private void registerConjunctionAndInitialize(ResolvableConjunction conjunction) {
+        if (!estimators.containsKey(conjunction)) {
+            estimators.put(conjunction, new ConjunctionAnswerCountEstimator(this, conjunction));
+        }
+
+        if (estimators.get(conjunction).mayInitialize()) {
+            initializationStack.add(conjunction);
+            estimators.get(conjunction).initialize();
+            assert initializationStack.get(initializationStack.size() - 1) == conjunction;
+            initializationStack.remove(initializationStack.size() - 1);
+
+            if (cycleHeads.size() == 1 && cycleHeads.contains(conjunction)) {
+                // Reset all cyclic paths
+                for (ResolvableConjunction conjunctionToReset : toReset) {
+                    estimators.get(conjunctionToReset).reset();
+                }
+            }
+        }
+    }
+
+    private void reportInitializationCycle(ResolvableConjunction conjunction) {
+        cycleHeads.add(conjunction);
+        for (int i = initializationStack.size() - 1; initializationStack.get(i) != conjunction; i--) {
+            toReset.add(initializationStack.get(i));
+        }
+    }
+
+    private static class ConjunctionAnswerCountEstimator {
+
+        private final ResolvableConjunction conjunction;
+        private final AnswerCountEstimator answerCountEstimator;
+        private final LogicManager logicMgr;
+        private final AnswerCountModel answerCountModel;
+        private Map<Variable, LocalEstimate> unaryEstimateCover;
+        private Map<Resolvable<?>, List<LocalEstimate>> retrievableEstimates;
+        private Map<Resolvable<?>, LocalEstimate> inferrableEstimates;
+        private Map<Resolvable<?>, LocalEstimate> unaryEstimates;
+
+        private long fullAnswerCount;
+        private long negatedsCost;
+
+        private enum InitializationStatus {NOT_STARTED, RESET, IN_PROGRESS, COMPLETE}
```

Comment:
"initialisation" is not something that really has states normally. Something else usually has the state of "initialising" for example. I'd propose `EstimationStatus {EMPTY, RESET, PARTIAL, COMPLETE}` for example.

If using this naming, we further get hints that the `mayInitialise` can be renamed - maybe `hasEstimate`? Then, `initialise()` could be `computeEstimate()`?

Reply:
That would need me to rename the estimateAnswers methods to avoid ambiguity.

Currently, there's a (recursive) initialisation phase which collects the local estimates, and a query phase which uses these  local estimates to answer complex queries (without having to recurse).
`needsPreparation` and `prepareLocalEstimates` may be a better fit? `hasEstimate` is not the write set of conditions, because that would correspond to `state==COMPLETE`.